DatadogRUM/Sources/Instrumentation/AppHangs/AppHangsWatchdogThread.swift

```diff
@@ -0,0 +1,131 @@
+    ) {
+        self.appHangThreshold = appHangThreshold
+        self.idleInterval = appHangThreshold * Constants.tolerance
+        self.mainQueue = queue
+        self.dateProvider = dateProvider
+        self.telemetry = telemetry
+        super.init()
+        self.name = "com.datadoghq.app-hang-watchdog"
+    }
+
+    override func main() {
+        let mainThreadTask = self.mainThreadTask
+
+        while !isCancelled {
+            defer {
+                // Before continuing, sleep for a while, to reduce the CPU consumption by watchdog thread.
+                Thread.sleep(forTimeInterval: idleInterval)
+            }
+
+            let waitStart: DispatchTime = .now()
+
+            // Schedule task on the main thread to measure how fast it responds
+            mainQueue.async {
+                mainThreadTask.signal()
+            }
+
+            // Await task completion
+            let result = mainThreadTask.wait(timeout: waitStart + appHangThreshold)
+
+            if result == .success {
+                // This is predominant case of "no hang" situation. The main thread executed the task way below hang threshold.
+                continue
+            } // Otherwise, `result == .timeOut`, meaning that the main thread task is delayed to more than hang threshold.
+
+            // We expect to be here roughtly `~appHangThreshold` after `wait()`. If this code is executed well after (50% margin),
+            // assume that the thread was suspended and app just woke up. In such case, ignore the hang as likely false-positive.
+            if interval(from: waitStart, to: .now()) > appHangThreshold * 1.5 {
+                DD.logger.debug("Ignoring likely false-positive App Hang")
+                continue // ignore likely false-positive
+            }
+
+            // TODO: RUM-2925 Capture stack trace of the pending App Hang
+
+            // Previous wait timed out, so wait again for the task completion, this time infinitely until the hang ends.
+            mainThreadTask.wait()
+
+            // The hang has finished.
+            let hangDuration = interval(from: waitStart, to: .now())
+
+            if hangDuration > 30 { // sanity-check
+                // If the hang duration is irrealistically long (way more than assumed iOS watchdog termination limit: ~10s), send telemetry.
+                // This could be another false-positive caused by thread suspension between the two `wait()` calls.
+                telemetry.debug("Detected an App Hang with an unusually long duration", attributes: ["hang_duration": hangDuration])
+                return
+            }
+
+            let appHang = AppHang(
+                date: dateProvider.now,
+                duration: hangDuration
+            )
+            onHangEnded?(appHang)
+        }
+    }
+
+    private func interval(from t1: DispatchTime, to t2: DispatchTime) -> TimeInterval {
+        TimeInterval(t2.uptimeNanoseconds - t1.uptimeNanoseconds) / 1_000_000_000
+    }
```

Comment:
I prefer adding convenience APIs after the time - if we see this pattern is repeated over and over, then it could be a moment for refactor and adding convenience.